release/models/interfaces/openconfig-if-aggregate.yang

```diff
@@ -204,12 +213,63 @@ module openconfig-if-aggregate {
       type leafref {
         path "/oc-if:interfaces/oc-if:interface/oc-if:name";
       }
+      status deprecated;
       description
         "Specify the logical aggregate interface to which
         this interface belongs";
     }
   }
 
+  grouping aggregation-ethernet-config {
+    description
+      "Adds configuration items for dedicated aggregation container
+      for Ethernet interfaces belonging to a logical aggregate / LAG";
+
+    leaf aggregate-id {
+      type leafref {
+        path "/oc-if:interfaces/oc-if:interface/oc-if:name";
+      }
+      description
+        "Specify the logical aggregate interface to which
+        this interface belongs";
+    }
+
+    leaf port-priority {
+      type uint16;
+      default "32768";
+      description
+        "This interface's priority in its aggregate interface";
+    }
+  }
+
+  grouping aggregation-ethernet-top {
+    description
+      "Top level configuration and state data for Ethernet interface's
+      aggregation parameters";
+
+    container aggregation {
+      description
+        "Configuration and operational state data for Ethernet interface's
+        aggregation parameters";
+
+      container config {
+        description
+          "Configuration for interface's aggregation parameters";
+
+        uses aggregation-ethernet-config;
+      }
+
+      container state {
+        config false;
+        description
+          "Operational state for interface's aggregation parameters";
+
+
+        uses aggregation-ethernet-config;
```

Comment:
There can be protocol agnostic aggregation parameters then there is also LACP specific.  For what is being discussed here, we are talking about LACP so would likely propose either a LACP container or prefixing of nodes.  Container organization is likely cleaner/expansible

Reply:
So should I just rename the new container at `/interfaces/interface/ethernet` from `aggregation` to `lacp`? Would the leaf `aggregate-id` still belong there then?

Reply:
IMHO, if this is LACP specific, then the better change is to have this be stored under `/lacp`. Our pattern is that the hierarchy is protocol-centric, and interfaces don't become a collecting ground for all parameters.

We'll have to fix making sure that the `config false` is applied at the right point in the tree. @dplore -- this will need us to discuss whether this is a backwards incompatible change (making a node that was `config false` now `config true` but not changing the `config` state of any particular leaf).

If there are generic statistics to do with aggregation, then `/interfaces/interface/aggregation` would be preferred -- but I'm not sure that I'm immediately aware of any.

Reply:
My understanding is that `/lacp/interfaces/interface/config` refers to aggregate interfaces, whereas LACP data for individual Ethernet member interfaces is under `/lacp/interface/interfaces/members/member`. There is currently no config container under the latter, that is why I initially proposed to add the container `/lacp/interface/interfaces/members/member/config`.

@dplore, do you specifically mean that we should consider adding `port-priority` under `/lacp/interface/interface/config`? Because I'm not sure what the interpretation of this container would be then, since `port-priority` is Ethernet interface-specific, while `/lacp/interface/interface/config` refers to aggregate interfaces.

So for example, if we had aggregate interface `Port-Channel1` with `Ethernet1` and `Ethernet2` as member interfaces, `Port-Channel1` would appear  as a key under `/lacp/interfaces/interface`, whereas `Ethernet1` and `Ethernet2` would appear as keys under `/lacp/interfaces/interface/members/member`. So if I wanted to give `Ethernet1` a lower priority within `Port-Channel1`, I'm not sure how I would implement that with `port-priority` being a leaf in `/lacp/interfaces/interface/config`, since that container can configure `Port-Channel1` but not `Ethernet1`.

Reply:
I think it does need to be `/lacp/interfaces/interface/members/member/config` given that its per member configuration. This was the `config true` container I was suggesting we would add.

Reply:
Ok great! I noticed that there are two `config false`: one under `/lacp/interfaces/interface/members` and one under `/lacp/interfaces/interface/members/member/state`:
```
    container members {
      config false;
      [...]

      list member {
        key "interface";
        [...]

        leaf interface {
          [...]
        }

        container state {

          config false;
```

So in order to create the new container `/lacp/interfaces/interface/members/member/config` with `config true`, the line `config false` from `/lacp/interfaces/interface/members` would need to be removed. Changing an existing container from `config false` to `config true` would make this a breaking change, right?

Reply:
Yes - it would be backwards incompatible but likely varied/minimal impact to existing implementations imo

I guess I'll offer another viewpoint for consideration back to the `/interfaces` subtree.  It's really a design pattern/view at the end of the day

Now, _most_ implementations that support link aggregation will usually have the configuration reference construct to the parent virtual interface at the member

* IOS-XR: https://www.cisco.com/c/en/us/td/docs/iosxr/ncs5xx/l2vpn/66x/b-l2vpn-cg-66x-ncs540/b-l2vpn-cg-63x-ncs5xx_chapter_0100.html
* EOS: https://www.arista.com/en/um-eos/eos-port-channels-and-lacp
* JUNOS: https://www.juniper.net/documentation/us/en/software/junos/interfaces-security-devices/topics/topic-map/security-interface-config-aggregated-ethernet.html
* Linux bonding w/ iproute2: http://www.uni-koeln.de/~pbogusze/posts/LACP_configuration_using_iproute2.html

But the opposite is possible and very well a mix depending on bootstrap mechanisms w/ Linux bonding or teaming (conf files, init declarations, etc..)

Now, the `ethernet` container under `interfaces` is probably not the right anchor point as @robshakir pointed out as this ties aggregation parameters specifically to ethernet and there are other possibilities for other link types albeit more and more rare these days - especially in relation to OpenConfig (where its likely many other features/subtrees are non-existant for the use-cases today)

But I think the `aggregation` container would very well be a good anchor imo where you can have protocol agnostic (or other) coupled with a sub-container for `lacp` specific that could be gated by a `when ...` back to the interface `type`

This also provides a known consistent location to target for members (`/interfaces/interfaces/aggregation`).  If we add other LAG parameters in the future they could be dispersed into other subtrees for protocol agnostic or protocol based.

It would however create the split for state as it is today.... much like viewing LACP related state under `show interfaces...` vs. something like `show lacp interfaces ...` - I think this is also a common split among implementations where the latter provides the necessary state detail and the prior only provides a reference to the parent

Reply:
So we have `/lacp/interfaces/interface/members/member/config` as a new container to be added (config true) for providing lacp configuration which is per-interface.    

This PR should also deprecate existing /interfaces/interface/ethernet/(config|state)/aggregate-id as originally proposed and add aggregate-id in the new container  as `/lacp/interfaces/interface/members/member/config/aggregate-id`.

My interpretation is does not need to be treated as  a breaking change because this is adding a new container, not changing the attributes of an existing container.  In other words, the existing `/lacp/interfaces/interface/members` container can stay as config false, I think?

The existing `/interfaces/interface/aggregation/config/` container remains for protocol agnostic link aggregation configuration which is applied to the 'virtual', aggregated interface.

Reply:
It's backwards incompatible due to the following which is inherited to child nodes

```
  grouping aggregation-lacp-members-top {
    description
      "Top-level grouping for aggregate members list";

    container members {
      config false;
      description
        "Enclosing container for the list of members interfaces of
        the aggregate. This list is considered operational state
        only so is labeled config false and has no config container";
```

But as I mentioned, highly unlikely to be a big issue among implementations.  the `config false` needs removal at the container first.

Reply:
Can the container `members` with list `member` be config false while a child container `members/member/config` is config true?  I don't see this prohibited in RFC6020.

Reply:
> Can the container `members` with list `member` be config false while a child container `members/member/config` is config true? I don't see this prohibited in RFC6020.

No - any `config false` at a hierarchy means all subhierarchies inherit and cannot be overridden back to `config true`

From RFC 6020/7950

```
When a node is tagged with "config false", its subhierarchy is flagged as state data.
```

Compilation errors from pyang/yanglint:

```
error: config true cannot be set when the parent is config false

Configuration node cannot be child of any state data node. (Path "/a:a/b/config".)
```

The RFC verbiage could maybe be a bit more verbose but if you consider state only subtrees where some configuration hierarchy would be nested deep into the tree, this would conflict w/ many internal behaviors that would not permit writing of the parent object/node (in this case the members container, member list, key, etc..)

In addition - the list key is also currently a leafref to the state node and would need to change to the config node if proceeding w/ this change

```
           +--ro members
              +--ro member* [interface]
                 +--ro interface    -> ../state/interface
```

Reply:
Ok, thanks everyone for the additional details. I am almost ready to submit a new version of the PR, adding the new `/lacp/interfaces/interface/members/member/config`, but I still have some questions remaining. A summary of my intended changes:

1. Add the `config` container with the following leaves:
```
module: openconfig-lacp
  +--rw lacp
     +--rw interfaces
        +--rw interface* [name]
           +--rw members
              +--rw member* [interface]
                 +--rw config
                    +--rw interface       oc-if:base-interface-ref
                    +--rw port-priority   uint16
```
2. Remove the `config false` from `/lacp/interfaces/interface/members`. This is necessary because otherwise, as @earies pointed out, the child container `/lacp/interfaces/interface/members/member/config` cannot be `config true`.
3. Change the key to list `/lacp/interfaces/interface/members/member` from `state/interface` to `config/interface`:
```
  +--rw lacp
     +--rw interfaces
        +--rw interface* [name]
           +--rw members
              +--rw member* [interface]
-                 +--ro interface    -> ../state/interface
+                 +--rw interface    -> ../config/interface
```
4. Deprecate `/interfaces/interface/ethernet/(config|state)/aggregate-id` and move it to the new `/lacp/interfaces/interface/members/member/(config|state)`. It’s on this last point that I still have some questions:

      a. What is the meaning of `aggregate-id` at this new path? Because the information about the aggregate interface is already contained in the path. For example if I, want to add `Ethernet1` to `Port-Channel1`, I would run something like`update /lacp/interfaces/interface[name=Port-Channel1]/members/member[interface=Ethernet1]/config/interface “Ethernet1”`. So in this case the the value of `/lacp/interfaces/interface[name=Port-Channel1]/members/member[interface=Ethernet1/config/aggregate-id` would be `Port-Channel1`.
      
      So then should `aggregate-id` be a leafref pointing to `/lacp/interfaces/interface/name`?

      b. With the deprecation of the leaf `/interfaces/interface/ethernet/(config|state)/aggregate-id`, how will an Ethernet interface be added to a `static` (non-LACP) aggregate interface? The new location of the leaf under the `/lacp` container does not seem like the right place for config that is explicitly non-LACP.

Reply:
Added a new commit reverting my current changes and adding instead `/lacp/interfaces/interface/members/member/config` with `config true`. Haven't yet made any changes (deprecate or move) to `/interfaces/interface/ethernet/(config|state)/aggregate-id` -- will wait for someone to confirm that it's okay to do. I'm still concerned about how moving that attribute to `/lacp/interfaces/interface/members/member/(config|state)` will affect configuring static/non-LACP aggregate interfaces